base/tps/src/main/java/org/dogtagpki/server/tps/rest/TPSCertService.java

```diff
@@ -162,6 +161,52 @@ public Response findCerts(String filter, String tokenID, Integer start, Integer
 
             return createOKResponse(response);
 
+        } catch (Exception e) {
+            throw new PKIException(e.getMessage());
+        }
+    }
+
+    private Response findAllCerts(String filter, Integer start, Integer size) {
+        String method = "TPSCertService:findAllCerts: ";
+        String msg = "";
+        TPSCertCollection response = new TPSCertCollection();
+
+        logger.debug("TPSCertService.findAllCerts({}, {}, {})", filter, start, size);
+
+        TPSEngine engine = TPSEngine.getInstance();
+        try {
+            TPSSubsystem subsystem = (TPSSubsystem) engine.getSubsystem(TPSSubsystem.ID);
+            TPSCertDatabase certDatabase = subsystem.getCertDatabase();
+            Iterator<TPSCertRecord> certRecs = certDatabase.findRecords(filter).iterator();
+            TokenDatabase tokenDatabase = subsystem.getTokenDatabase();
+            List<String> authorizedProfiles = getAuthorizedProfiles();
+            if (authorizedProfiles == null) {
+                msg = "authorizedProfiles null";
+                logger.debug("{}{}", method, msg);
+                throw new PKIException(method + msg);
+            }
+
+            int i = 0;
+            // skip to the start of the page
+            for (; i < start && certRecs.hasNext(); i++) certRecs.next();
+
+            if (certRecs.hasNext()) {
+                certRecs.forEachRemaining(certRecord -> {
+                    int j = 0;
+                    String tokenID = certRecord.getTokenID();
+                    TokenRecord tokenRecord = null;
+                    try {
+                        tokenRecord = tokenDatabase.getRecord(tokenID);
+                    } catch (Exception e) {
+                        // Proceed to next token if this one doesn't exist.
+                    }
+                    if (tokenRecord != null && authorizedProfiles.contains(UserResource.ALL_PROFILES) && j < size) {
```

Comment:
I think we need to check the record type too:
```
if (tokenRecord == null) continue;

if (j >= size) continue; // outside the current page

String type = record.getType();
if (!authorizedProfiles.contains(UserResource.ALL_PROFILES) && !authorizedProfiles.contains(type))
    continue; // no access

response.addEntry(createCertData(certRecord));
```

Reply:
Ack, I added that in, it got lost somehow. You can't use break/continue etc with `forEachRemaining`, so there is a limitation there. I'm not entirely convinced this is the right mechanism.